demos/https/iot_demo_https_s3_download_async.c

```diff
@@ -0,0 +1,741 @@
+    httpsClientStatus = IotHttpsClient_GetUrlPath( IOT_DEMO_HTTPS_PRESIGNED_GET_URL, (size_t)strlen( IOT_DEMO_HTTPS_PRESIGNED_GET_URL ), &pPath, &pathLen);
+    if (httpsClientStatus != IOT_HTTPS_OK)
+    {
+        IotLogError("An error occurred in IotHttpsClient_GetUrlPath() with error code %d.", httpsClientStatus);
+        return EXIT_FAILURE;
+    }
+
+    /* Retrieve the address location and length from the IOT_DEMO_HTTPS_PRESIGNED_GET_URL. */
+    httpsClientStatus = IotHttpsClient_GetUrlAddress( IOT_DEMO_HTTPS_PRESIGNED_GET_URL, (size_t)strlen( IOT_DEMO_HTTPS_PRESIGNED_GET_URL ), &pAddress, &addressLen);
+    if (httpsClientStatus != IOT_HTTPS_OK)
+    {
+        IotLogError("An error occurred in IotHttpsClient_GetUrlAddress() with error code %d.", httpsClientStatus );
+        return EXIT_FAILURE;
+    }
+
+    /* Set the connection configurations. */
+    connConfig.pAddress = pAddress;
+    connConfig.addressLen = addressLen;
+    connConfig.port = IOT_DEMO_HTTPS_PORT;
+    /* We disable SNI here because the address specified includes the S3 bucket name. */
+    connConfig.flags |= IOT_HTTPS_DISABLE_SNI; /* Disable SNI, enable TLS, enable Persistent connections, is HTTP/1.1 */
+    connConfig.pCaCert = IOT_DEMO_HTTPS_TRUSTED_ROOT_CA;
+    connConfig.caCertLen = sizeof( IOT_DEMO_HTTPS_TRUSTED_ROOT_CA );
+    connConfig.userBuffer.pBuffer = _pConnUserBuffer;
+    connConfig.userBuffer.bufferLen = sizeof(_pConnUserBuffer);
+    connConfig.pClientCert = ( ( IotNetworkCredentials_t* )pNetworkCredentialInfo )->pClientCert;
+    connConfig.clientCertLen = ( ( IotNetworkCredentials_t* )pNetworkCredentialInfo )->clientCertSize;
+    connConfig.pPrivateKey = ( ( IotNetworkCredentials_t* )pNetworkCredentialInfo )->pPrivateKey;
+    connConfig.privateKeyLen = ( ( IotNetworkCredentials_t* )pNetworkCredentialInfo )->privateKeySize;
+    connConfig.pNetworkInterface = pNetworkInterface;
+
+    /* Set the configurations needed for an asynchronous request. */
+    asyncInfo.callbacks.appendHeaderCallback = _appendHeaderCallback;
+    asyncInfo.callbacks.readReadyCallback = _readReadyCallback;
+    asyncInfo.callbacks.responseCompleteCallback = _responseCompleteCallback;
+    asyncInfo.callbacks.connectionClosedCallback = _connectionClosedCallback;
+    asyncInfo.callbacks.errorCallback = _errorCallback;
+    
+    /* Create the mutex to protect the pool of requests. */
+    if( IotMutex_Create( &( _inUseRequestsMutex ), false ) == false)
+    {
+        IotLogError("Failed to create a mutex to protect the request pool.");
+        return EXIT_FAILURE;
+    }
+    
+    /* Create the semaphore for waiting for the response to finish. */
+    if( IotSemaphore_Create( &( _fileFinishedSem ), 0, 1 ) == false )
+    {
+        IotLogError( "Failed to create a semaphore to wait for the response to finish." );
+        IotMutex_Destroy( &( _inUseRequestsMutex ) );
+        return EXIT_FAILURE;
```

Comment:
Please use cleanup macro here and on line 588.

Reply:
OK.